racket-src/body/type-check-goal.rkt

```diff
@@ -135,9 +135,43 @@
                            (&& ((Ty_op <= Ty_field) ...)))
    ]
 
+  [; type-check ReifyFnPointer casts
+   (type-of/Operand Γ Operand Ty_fn)
+   (fn-ty-signature Γ Ty_fn (∀ KindedVarIds (implies Biformulas ((Ty_arg ...) -> Ty_ret))))
+   (where/error number_args ,(length (term (Ty_arg ...))))
+   (where/error Ty_fnptr (rigid-ty (fn-ptr "Rust" number_args) (Ty_arg ... Ty_ret)))
+   (where/error Ty_reified (∀ KindedVarIds (implies Biformulas Ty_fnptr)))
+   ----------------------------------------
+   (type-check-goal/Rvalue Γ
+                           (cast Operand as Ty_target)
```

Comment:
OK, so, `Ty_target` should *always* be a fn pointer type, right? I think I expect this to look a bit more like this

```
[(is-fn-pointer Ty_target) ; some kind of simple check
 (type-of/Operand Γ Operand Ty_fn)
 (fn-cast-goal Γ Ty_fn Ty_target Goal)
 ----------------------------------------
 (type-check-goal/Rvalue Γ (cast Operand as Ty_target) Goal)
 ]
```

where `fn-cast-goal` has a recursion shape like this

```
[(fn-cast-goal Γ Ty Ty_target)
 ----------------------------------------
 (fn-cast-goal Γ (∀ KindedVarIds Ty) Ty_target (∀ KindedVarIds Goal))

[(fn-cast-goal Γ Ty Ty_target)
 ----------------------------------------
 (fn-cast-goal Γ (implies Biformulas Ty) Ty_target (implies Biformulas Goal))

[(where/error (fn FnId KindedVarIds_fn (Ty_arg ...) -> Ty_ret _ _ _) (find-fn Γ FnId))
 (where/error Substitution (create-substitution KindedVarIds_fn Parameters))
 (where/error (Ty_argsubst ...) ((apply-substitution Substitution Ty_arg) ...))
 (where/error Ty_retsubst (apply-substitution Substitution Ty_ret))
 (where/error Ty_fnptr (rigid-ty (fn-ptr "Rust" number_args) [Ty_argsubst ... Ty_retsubst]))
 ----------------------------------------
 (fn-cast-goal Γ (rigid-ty (fn-def FnId) Parameters) Ty_target (Ty_fnptr <= Ty_target))
```

This way we aren't hardcoding the type structure so much, we just translate it to goals as we pass through it.

Reply:
I moved the binders of the operand's type `Ty_fn` to the goal. Considering that `Ty_target` can have its own `∀` binders, does this give the correct behavior? Regarding not hardcoding the type structure, I made some changes to `fn-ty-signature`. Does that already accomplish what you had in mind?

Reply:
> Considering that Ty_target can have its own ∀ binders, does this give the correct behavior? 

Hmm, I thought so, but now I'm realizing maybe this wasn't quite right. What we want may actually be more like

```
(reified-fn-type Gamma Ty_in Ty_out)
```

...which does the translation, still I think in the recursive way I described, but changing the innermost `fndef` to `fnptr`, and then...

```
(reified-fn-type Gamma Ty_source Ty_reified)
(where/error Goal (Ty_reified <= Ty_target))
```

Reply:
In particular, the structure I described proves that the cast works for *every possible value* of the input lifetimes, but actually we only need it to hold for *some value*. The code as I wrote it should error out, for example, if you do this cast

```rust
fn foo(x: &u8) {}

let x: fn(&'static u8) = foo;
```

(you could try adding a test)

but the version above would work fine (since it would reify `foo` to `fn(&u8)`, and then this is a subtype of `fn(&'static u8)`).